components/ntp_background_images/browser/ntp_sponsored_images_data.cc

```diff
@@ -196,83 +135,230 @@ void NTPSponsoredImagesData::ParseCampaignsList(
+      if (condition_matchers) {
+        // Condition matchers are optional.
         for (const auto& condition_matcher : *condition_matchers) {
-          const auto& dict = condition_matcher.GetDict();
-          const auto* const pref_path =
-              dict.FindString(kWallpaperConditionMatcherPrefPathKey);
-          if (!pref_path) {
-            continue;
+          const base::Value::Dict* const condition_matcher_dict =
+              condition_matcher.GetIfDict();
+          if (!condition_matcher_dict) {
+            // Invalid condition matcher.
+            return std::nullopt;
           }
 
-          const auto* const condition =
-              dict.FindString(kWallpaperConditionMatcherKey);
+          const std::string* const condition =
+              condition_matcher_dict->FindString(
+                  kCreativeConditionMatcherConditionKey);
           if (!condition) {
-            continue;
+            // Condition is required.
+            return std::nullopt;
+          }
+
+          const std::string* const pref_path =
+              condition_matcher_dict->FindString(
+                  kCreativeConditionMatcherPrefPathKey);
+          if (!pref_path) {
+            // Pref path is required.
+            return std::nullopt;
           }
 
           background.condition_matchers.emplace(*pref_path, *condition);
         }
       }
 
-      if (auto* viewbox = wallpaper.FindDict(kViewboxKey)) {
-        gfx::Rect rect(viewbox->FindInt(kXKey).value_or(0),
-                       viewbox->FindInt(kYKey).value_or(0),
-                       viewbox->FindInt(kWidthKey).value_or(0),
-                       viewbox->FindInt(kHeightKey).value_or(0));
-        background.viewbox.emplace(rect);
-      }
-      if (auto* background_color = wallpaper.FindString(kBackgroundColorKey)) {
-        background.background_color = *background_color;
+      // Wallpaper.
+      const base::Value::Dict* const wallpaper =
+          creative_dict->FindDict(kWallpaperKey);
+      if (!wallpaper) {
+        // Wallpaper is required.
+        return std::nullopt;
       }
-      if (auto* creative_instance_id =
-              wallpaper.FindString(kCreativeInstanceIDKey)) {
-        background.creative_instance_id = *creative_instance_id;
+
+      const std::string* const wallpaper_type =
+          wallpaper->FindString(kWallpaperTypeKey);
+      if (!wallpaper_type) {
+        // Wallpaper type is required.
+        return std::nullopt;
       }
-      if (auto* wallpaper_logo = wallpaper.FindDict(kLogoKey)) {
-        background.logo =
-            GetLogoFromValue(installed_dir, url_prefix, *wallpaper_logo);
+
+      if (*wallpaper_type == kImageWallpaperType) {
+        // Image.
+        background.wallpaper_type = WallpaperType::kImage;
+
+        const std::string* const relative_url =
+            wallpaper->FindString(kImageWallpaperRelativeUrlKey);
+        if (!relative_url) {
```

Comment:
do any of the relative URLs in this file need to check for path traversal?

Reply:
They do not, because they will be checked in `MaybeGetFilePathForRequestPath`. Thanks

Reply:
However, lets reject here too, so that imageNTT (unrelated to this PR) also gets that fix.

Reply:
Fixed